security/sensorhub-security-oauth/src/main/java/org/sensorhub/impl/security/oauth/OAuthAuthenticator.java

```diff
@@ -227,23 +231,73 @@ public Authentication validateRequest(ServletRequest req, ServletResponse resp,
+            if (accessToken == null && authHeader != null && !request.getServletPath().equals("/admin")) {
+                try {
+                    var credentials = parseBasicAuth(authHeader);
+                    if (credentials == null) {
+                        var clientCredentialsToken = parseBearerToken(authHeader);
+
+                        if (clientCredentialsToken != null) {
+                            // if client credentials token, assume it's a JWT token that contains needed info
+                            log.debug("Client Credentials Token = {}", clientCredentialsToken);
+
+                            String[] chunks = clientCredentialsToken.split("\\.");
+
+                            Base64.Decoder decoder = Base64.getUrlDecoder();
+
+                            String header = new String(decoder.decode(chunks[0]));
+
+                            log.debug("Client Credentials Token Header = {}", header);
+
+                            JsonElement headerData = JsonParser.parseString(header);
 
-                clientCredentialsToken = parseBearerToken(authHeader);
+                            JwkProvider jwkProvider =
+                                    new JwkProviderBuilder(new URL(config.bearerTokenConfig.jwksUri))
+                                            .cached(config.bearerTokenConfig.cacheSize,
+                                                    Duration.of(config.bearerTokenConfig.cacheDuration, ChronoUnit.MINUTES))
+                                            .build();
 
-                // if no OAuth token received check if credentials are provided in authorization header
-                // and use back channel direct auth flow
-                if (clientCredentialsToken == null && accessToken == null) {
+                            DecodedJWT decodedJWT;
+                            try {
+
+                                RSAPublicKey publicKey = (RSAPublicKey) jwkProvider.get(parseKidFromJson(headerData)).getPublicKey();
+                                Algorithm algorithm = Algorithm.RSA256(publicKey);
+
+                                JWTVerifier verifier = JWT.require(algorithm)
+                                        // specify any specific claim validations
+                                        .withIssuer(config.bearerTokenConfig.issuer)
+                                        .withAudience(config.bearerTokenConfig.audience)
+                                        .build();
 
-                    try {
+                                decodedJWT = verifier.verify(clientCredentialsToken);
 
-                        var credentials = parseBasicAuth(authHeader);
+                                String payload = new String(decoder.decode(decodedJWT.getPayload()));
 
-                        if (credentials == null) {
+                                log.debug("Client Token Payload = {}", payload);
+
+                                userInfo = JsonParser.parseString(payload);
```

Comment:
I don't think there is a need to parse `userInfo` here.
Just set `idToken=payload` and the logic down below should parse the user info just like when it's KC auth server sending it back.

Reply:
Refactored code to only validate the header according to JWT guidance for minimum token validation